sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/InsecureDeserialization.cs

```diff
@@ -0,0 +1,339 @@
+    {
+        public MultipleConstructorsNoConditionals()
+        {
+        }
+
+        public MultipleConstructorsNoConditionals(bool condition, string name)
+        {
+        }
+
+        protected MultipleConstructorsNoConditionals(SerializationInfo info, StreamingContext context)
+        {
+        }
+
+        void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context) { }
+    }
+
+    [Serializable]
+    public class WithConditionsISerializable : ISerializable
+    {
+        public string Url { get; }
+
+        public WithConditionsISerializable(string url)
+        {
+            if (string.IsNullOrEmpty(url))
+            {
+                Url = url;
+            }
+        }
+
+        protected WithConditionsISerializable(SerializationInfo info, StreamingContext context)
+        {
+            foreach (var item in info)
+            {
+                if (item.Name == "Url")
+                {
+                    Url = (string) item.Value;
+                }
+            }
+        }
+
+        void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context) { }
+    }
+
+    [Serializable]
+    public class WithDifferentConditionsISerializable : ISerializable
+    {
+        public string FirstName { get; }
+
+        public string LastName { get; }
+
+        public WithDifferentConditionsISerializable(string firstName, string lastName) // Compliant - FN: ctor has conditions on different args/fields/properties than deserialization constructor
+        {
+            FirstName = firstName ?? "";
+            LastName = lastName;
+        }
+
+        protected WithDifferentConditionsISerializable(SerializationInfo info, StreamingContext context)
+        {
+            FirstName = info.GetString("FirstName");
+            LastName = info.GetString("LastName") ?? string.Empty;
+        }
+
+        void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context) { }
+    }
+
+    [Serializable]
+    public class WithoutConditionsInDeserializationCtor : ISerializable
+    {
+        public string Url { get; }
+
+        public WithoutConditionsInDeserializationCtor(string url) // Noncompliant
+        {
+            if (string.IsNullOrEmpty(url))
+            {
+                Url = url;
+            }
+        }
+
+        protected WithoutConditionsInDeserializationCtor(SerializationInfo info, StreamingContext context)
+        {
+            Url = info.GetString("Url");
+        }
+
+        void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context) { }
+    }
+
+    [Serializable]
+    public class NoConstructor : IDeserializationCallback
+    {
+        public void OnDeserialization(object sender)
+        {
+        }
+    }
+
+    [Serializable]
+    public class WithoutConditionsInCtor : IDeserializationCallback
+    {
+        private string Name { get; }
+
+        public WithoutConditionsInCtor()
```

Comment:
If we add a parameter `string name` and use `Name = name`, there will be a FP (see following comment)